hydra-node/src/Hydra/Chain/Direct/Tx.hs

```diff
@@ -176,56 +178,54 @@ withOutputs newOutputs txbody =
 
 -- | Create the init transaction from some 'HeadParameters' and a single TxIn
 -- which will be used as unique parameter for minting NFTs.
+--
+-- TODO: Remove all the qualified 'Api' once the refactoring is over and there's
+-- no more import clash with 'cardano-ledger'.
+--
+-- TODO: Get rid of Ledger types in the signature and fully rely on Cardano.Api
 initTx ::
   -- | Participant's cardano public keys.
   [VerificationKey PaymentKey] ->
   HeadParameters ->
   TxIn StandardCrypto ->
   ValidatedTx Era
 initTx cardanoKeys HeadParameters{contestationPeriod, parties} txIn =
-  emptyTx
-    & withBody body
-    & withDatums dats
+  Api.toLedgerTx $
+    Api.unsafeBuildTransaction $
+      Api.emptyTxBody
+        & Api.addVkInputs [Api.fromLedgerTxIn txIn]
+        & Api.addOutputs (headOut : map mkInitial cardanoKeys)
  where
-  body =
-    emptyTxBody
-      & withInputs [txIn]
-      & withOutputs (headOut : initials)
-
-  dats = headDatum : [initialDatum vkey | vkey <- cardanoKeys]
-
-  headOut = TxOut headAddress headValue (SJust headDatumHash)
-
-  headAddress = scriptAddr $ plutusScript $ MockHead.validatorScript policyId
-
-  -- REVIEW(SN): how much to store here / minUtxoValue / depending on assets?
-  headValue = inject (Coin 2000000)
-
-  headDatumHash = hashData @Era headDatum
-
-  headDatum =
-    Data . toData $
-      MockHead.Initial
-        (contestationPeriodFromDiffTime contestationPeriod)
-        (map (partyFromVerKey . vkey) parties)
-
-  initials = map mkInitial cardanoKeys
-
-  mkInitial = TxOut @Era initialAddress initialValue . SJust . initialDatumHash
-
-  initialAddress = scriptAddr $ plutusScript MockInitial.validatorScript
-
-  -- TODO: should really be the minted PTs plus some ADA to make the ledger happy
-  initialValue = headValue
-
-  initialDatumHash = hashData @Era . initialDatum
-
-  initialDatum vkey = Data . toData $ MockInitial.datum $ pubKeyHash vkey
+  headOut =
+    Api.TxOut headAddress headValue headDatum
+   where
+    headScript = (Api.fromPlutusScript $ MockHead.validatorScript policyId)
+    headAddress = Api.mkScriptAddress networkId headScript
+    headValue = Api.lovelaceToTxOutValue $ Api.Lovelace 2_000_000
+    headDatum =
+      Api.mkTxOutDatum $
+        MockHead.Initial
+          (contestationPeriodFromDiffTime contestationPeriod)
+          (map (partyFromVerKey . vkey) parties)
+
+  mkInitial (Api.toPlutusKeyHash . Api.verificationKeyHash -> vkh) =
+    Api.TxOut initialAddress initialValue initialDatum
+   where
```

Comment:
Should: Remove this `where`. It is not required and the definitions below could live just fine in the same `where` as `mkInitial`. I'd argue having less indentation / scopes is better to read/modify.

Reply:
I would disagree! Nesting `where` is one of the best feature of the language IMO as it helps keeping things tidy and showing that some bits of code are only related to some other bits and not others. As soon as a function starts having more than 5 `where` or so, it becomes hard to keep track of what is relevant when. Grouping `where` clauses as such is helping readability and brain load a lot!